### PR TITLE
New addon: block cherry-picking

### LINKED ISSUE #2338: Holding right click will only drag that block
> Idea: holding right click instead of left click and dragging a block out of a script will only take out that block and not everything underneath it.

https://github.com/LLK/scratch-blocks/blob/develop/core/block_svg.js#L745 might be useful but who knows...

### PR DESCRIPTION
**Pull request by lisa-wolfgang**
Resolves #2338

### Changes

When the addon is enabled and the Ctrl key (or Cmd key on Macs) is held down, blocks that are in the middle of scripts can be dragged out without the blocks beneath it being pulled out as well. (This does not work on the tops of scripts.)

Before:

https://user-images.githubusercontent.com/43426138/132908195-220330e8-d483-4f37-be7a-ee51fbfab336.mp4

After:

https://user-images.githubusercontent.com/43426138/132907076-d7b20cf8-7b01-4375-87da-f89d53248f6a.mp4

There is also an option that enables cherry-picking on regular dragging:

image

### Reason for changes

Normally, dragging out a block from the middle of a script takes 3 gestures -- one to detach the bottom scripts, one to drag out the middle block, and one to replace the bottom scripts. This addon makes this entire process much simpler.

### Tests

Tested on Chromium 93.

### CODE REVIEW

**Comment by lisa-wolfgang** [comment]
Also, this is my first independently-developed addon! 🥳

File: addons/block-cherry-picking/userscript.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  document.addEventListener("keydown", function (e) {
+    if (e.metaKey) {
+      ctrlKeyPressed = true;
+    }
+  });
+  document.addEventListener("keyup", function (e) {
+    if (!e.metaKey) {
+      ctrlKeyPressed = false;
+    }
+  });
```
Comment (GarboMuffin) [inline addons/block-cherry-picking/userscript.js:27]: any reason these can't be shortened to eg.?

```js
  document.addEventListener("keydown", function (e) {
    ctrlKeyPressed = e.ctrlKey || e.metaKey;
  });
```

File: addons/block-cherry-picking/userscript.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  // `opt_healStack` is a built-in option in scratch-blocks that enables cherry-picking behavior.
+  // All this function does is enable that built-in option for every block.
+  BlocklyInstance.BlockSvg.prototype.unplug = function (opt_healStack) {
+    if (ctrlKeyPressed != invertDrag) {
```
Comment (GarboMuffin) [inline addons/block-cherry-picking/userscript.js:32]: Doesn't really matter here but I think we generally prefer `!==`

I think you need something like `&& !addon.self.disabled` to make dynamicDisable work

File: addons/block-cherry-picking/addon.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    {
+      "type": "notice",
+      "text": "On macOS, use the Cmd key instead of the Ctrl key.",
+      "id": "macContextDisabled"
+    }
+  ],
+  "credits": [
+    {
+      "name": "lisa_wolfgang",
+      "link": "https://github.com/lisa-wolfgang"
```
Comment (GarboMuffin) [inline addons/block-cherry-picking/addon.json:14]: Your link in custom-block-shape points to your Scratch profile instead. Would probably be best to pick one and be consistent

**Comment by Secret-chest** [comment]
> Chromium 93

@lisa-wolfgang Did you switch to Linux? Nice!

**Comment by lisa-wolfgang** [comment]
> @lisa-wolfgang Did you switch to Linux? Nice!

Nope, just using a Chromium-based browser.

File: addons/block-cherry-picking/userscript.js (hunk 1 above)
Comment (lisa-wolfgang) [reply addons/block-cherry-picking/userscript.js:27]: I just copied it over from the Ctrl + click to run scripts addon -- I'll definitely condense that to what you're suggesting.

File: addons/block-cherry-picking/userscript.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+export default async function ({ addon, global, console }) {
+  const BlocklyInstance = await addon.tab.traps.getBlockly();
+  const originalObject = BlocklyInstance.Block.prototype.unplug;
+
+  // Necessary to detect the CTRL/CMD key
+  let ctrlKeyPressed = false;
+  document.addEventListener("keydown", function (e) {
+    ctrlKeyPressed = e.ctrlKey || e.metaKey;
+  });
```
Comment (WorldLanguages) [inline addons/block-cherry-picking/userscript.js:9]: Don't forget `keyup`.

File: addons/block-cherry-picking/addon.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  "credits": [
+    {
+      "name": "lisa_wolfgang",
+      "link": "https://scratch.mit.edu/users/lisa_wolfgang"
+    }
+  ],
+  "userscripts": [
+    {
+      "url": "userscript.js",
+      "matches": ["https://scratch.mit.edu/projects/*"]
```
Comment (WorldLanguages) [inline addons/block-cherry-picking/addon.json:20]: Use word matcher `"projects"`

**Comment by WorldLanguages** [comment]
I'm not so sure `Enable without holding Ctrl (hold Ctrl for regular drag)` is a good idea. I can already see some users posting on Bugs & Glitches.

**Comment by mxmou** [comment]
> I'm not so sure `Enable without holding Ctrl (hold Ctrl for regular drag)` is a good idea. I can already see some users posting on Bugs & Glitches.

I think it's useful - could a tooltip saying something similar to "hold Ctrl to move entire stack" be added to the block while it's being dragged if the setting is enabled? It would mean that pressing Ctrl after starting dragging needs to work too.

**Comment by lisa-wolfgang** [comment]
> 
> 
> > I'm not so sure `Enable without holding Ctrl (hold Ctrl for regular drag)` is a good idea. I can already see some users posting on Bugs & Glitches.
> 
> I think it's useful - could a tooltip saying something similar to "hold Ctrl to move entire stack" be added to the block while it's being dragged if the setting is enabled? It would mean that pressing Ctrl after starting dragging needs to work too.

Not sure about that -- the tooltip could seem more annoying than helpful to many users, and even without that factor, being able to switch in or out of cherry-picking mid-drag would be relatively complicated, feel a little janky to perform, and possibly break other things.

**Comment by lisa-wolfgang** [comment]
> I'm not so sure `Enable without holding Ctrl (hold Ctrl for regular drag)` is a good idea. I can already see some users posting on Bugs & Glitches.

@WorldLanguages Having users posting on Bugs & Glitches shouldn't be a reason for us to limit functionality. A better solution would be to implement that recently changed settings feature we were discussing.

**Comment by WorldLanguages** [comment]
Okay, let's keep the setting, but add a notice that explains in more details what it does.

**Comment by WorldLanguages** [comment]
We ended up with these after some discussion on Discord:
image

**Review by WorldLanguages — APPROVED** [review]
Merging soon

**Comment by lisa-wolfgang** [comment]
Lol we really need per-addon documentation

**Pull request merged**